Editorial review: Document scroll-triggered animations (#43158)
mdn/content: files/en-us/web/css/guides/animation_triggers/index.md

---
title: CSS animation triggers
short-title: Animation triggers
slug: Web/CSS/Guides/Animation_triggers
page-type: css-module
spec-urls: https://drafts.csswg.org/animation-triggers-1/
sidebar: cssref
---

The **CSS animation triggers** module provides functionality for triggering standard time-based [CSS animations](/en-US/docs/Web/CSS/Guides/Animations) when a particular trigger occurs, such as the user scrolling an element to a certain scroll offset inside the viewport, or a JavaScript event firing on an element.

## Animation triggers in action

> In the below example, start scrolling down. When the first "bouncer" text appears in the viewport, the ball will bounce off to the right. When the "bouncer" text scrolls off the top of the viewport, the ball animation will play in reverse, moving it back up to the top-left. Keep scrolling: when the "another bouncer" text appears in the viewport and scrolls off to the top, the same process will repeat.

Reviewer: Avoid "below".
The appearance and disappearance are triggers (the triggers also happen when scrolling up)
```suggestion
Scroll down and up the box of text. The appearance of the "bouncer" text will trigger the ball's animation. The "bouncer" text scrolling out of the viewport triggers a reverse animation. Keep scrolling: the same process will repeat when the "another bouncer" text appears in and then disappears from the viewport.
```

```html hidden live-sample___in-action
<section>
  <div>
    <p>scroll down</p>
    <p id="trigger">bouncer</p>
    <p>keep scrolling</p>
    <p>keep scrolling</p>
    <p>keep scrolling</p>
    <p id="trigger2">another bouncer</p>
    <p>scroll up</p>
  </div>
</section>

<span id="ball"><span></span></span>
```

```css hidden live-sample___in-action
#ball,
#ball span {
  animation:
    moveright 2s 1 ease-out both,
    moveright 2s 1 ease-out forwards;
  animation-trigger:
    --t play-forwards play-backwards,
    --t2 play-forwards play-backwards;
}
#ball span {
  animation-name: bounce, bounce;
}

#trigger {
  timeline-trigger-name: --t;
  timeline-trigger-source: view();
}
#trigger2 {
  timeline-trigger-name: --t2;
  timeline-trigger-source: view();
}

html {
  font-family: sans-serif;
}

* {
  box-sizing: border-box;
}

@layer scroller {
  section {
    border: solid;
    width: 200px;

>     height: 400px;

Reviewer: This makes it more obvious that the ball is not in the rectangle and makes it easier to scroll through (easier on limited mobiliity / carpal tunnel)

```suggestion
    height: 250px;
    margin-top: 50px;
```

  #ball {
    position: fixed;
    top: 96vh;
  }
  #ball span {
    background: red;
    border-radius: 50%;
    height: 5vw;
    display: block;
    aspect-ratio: 1/1;
  }

  @keyframes moveright {
    from {
      transform: translatex(0);
    }
    to {
      transform: translatex(90vw);
    }
  }

  @keyframes bounce {
    9%,
    24%,
    35%,
    44%,
    51%,
    58%,
    63%,
    68%,
    72%,
    76%,
    to {
      transform: translatey(0);
      animation-timing-function: ease-out;
    }
    from,
    17%,
    30%,
    40%,
    48%,
    55%,
    61%,
    66%,
    70%,
    74% {
      animation-timing-function: ease-in;
    }
    0% {
      transform: translatey(-96vh);
    }
    17% {
      transform: translatey(-57.6vh);
    }
    30% {
      transform: translatey(-34.56vh);
    }
    40% {
      transform: translatey(-20.74vh);
    }
    48% {
      transform: translatey(-12.44vh);
    }
    55% {
      transform: translatey(-7.46vh);
    }
    61% {
      transform: translatey(-4.48vh);
    }
    66% {
      transform: translatey(-2.69vh);
    }
    70% {
      transform: translatey(-1.61vh);
    }
    74% {
      transform: translatey(-0.97vh);
    }
  }

> }

Reviewer: add the "no support" code block here for browsers that don't support triggers yet or ever

Author: done

```

{{embedlivesample("in-action", "100%", 400)}}

> This example uses the {{cssxref("timeline-trigger-name")}} and {{cssxref("timeline-trigger-source")}} properties to create animaton triggers on the "bouncer" and "another bouncer" paragraphs. The bouncing ball has multiple {{cssxref("animation")}}s set on it, plus an {{cssxref("animation-trigger")}} property that references the trigger names set on the "bouncer" and "another bouncer" paragraphs and specifies animation actions to perform when the animations are triggered. This causes the ball's animations to play when the paragraphs appear in the viewport, and play in reverse when they scroll off the top of the viewport.

Reviewer: Simplifying down to "in action" levels of explanation (explaining what is occuring with less "code" detail, as the HTML and CSS aren't shown, but enough so they know what props they may want to click to). And removing "top" mention since it happens in both directions

```suggestion
Both trigger elements have a {{cssxref("timeline-trigger-name")}} and {{cssxref("timeline-trigger-source")}}, defining the text blocks as named scroll-view animation triggers. The bouncing ball has the bouncing {{cssxref("animation")}} set on it twice, plus an {{cssxref("animation-trigger")}} property that references both trigger names and specifies animation actions to perform whenever an animation is activated by a trigger coming into view and deactivated when the trigger exits the viewport.
```

  - : The {{cssxref("timeline-range-name")}} data type: Understanding the various timeline range names.

## Related concepts

- [CSS animations](/en-US/docs/Web/CSS/Guides/Animations) module
  - {{cssxref("animation-timeline")}}
  - {{cssxref("@keyframes")}} at-rule
  - [`<keyframe-selector>`](/en-US/docs/Web/CSS/Reference/Selectors/Keyframe_selectors)
- [CSS scroll-driven animations](/en-US/docs/Web/CSS/Guides/Scroll-driven_animations) module
  - {{cssxref("animation-range")}} shorthand
    - {{cssxref("animation-range-end")}}
    - {{cssxref("animation-range-start")}}
  - {{cssxref("scroll-timeline")}} shorthand
    - {{cssxref("scroll-timeline-axis")}}
    - {{cssxref("scroll-timeline-name")}}
  - {{cssxref("timeline-scope")}}
  - {{cssxref("view-timeline")}} shorthand
    - {{cssxref("view-timeline-axis")}}
    - {{cssxref("view-timeline-inset")}}
    - {{cssxref("view-timeline-name")}}

## Specifications

{{Specifications}}

## See also

- [Using CSS animations](/en-US/docs/Web/CSS/Guides/Animations/Using)
- [CSS scroll-driven animations](/en-US/docs/Web/CSS/Guides/Scroll-driven_animations)

> - [CSS scroll-triggered animations are coming!](https://developer.chrome.com/blog/scroll-triggered-animations) on developer.chrome.com (2025)

Reviewer: ```suggestion
- [CSS scroll-triggered animations are coming!](https://developer.chrome.com/blog/scroll-triggered-animations) via developer.chrome.com (2025)
```

Author: I still disagree with using the word "via" in these sentences. The article we are linking to is on that site. saying via makes it sound like we are going to that site, and then travelling to a different location to get to the linked article, which we're not.

Reviewer: we're all over the map on this. We should likely have a guide so we can be consistent.

by Person Name
via Name of Org
on website.com

Reviewer: Created https://github.com/mdn/content/issues/44119

---
title: CSS animation triggers
short-title: Animation triggers
slug: Web/CSS/Guides/Animation_triggers
page-type: css-module
spec-urls: https://drafts.csswg.org/animation-triggers-1/
sidebar: cssref
---

> The **CSS animation triggers** module provides functionality for triggering standard time-based [CSS animations](/en-US/docs/Web/CSS/Guides/Animations) when a particular trigger occurs, such as the user scrolling an element to a certain scroll offset inside the scrollport.

Reviewer: I think we should provide a bit more information: enough for someone landing on this page to understand what triggered animations are and how they related to CSS.. Let's provides a bit more of an overview of the module. a rough draft.... 

>By specifying a trigger for an animation, that animation becomes a triggered animation, delaying its play until the trigger occurs. These triggers can be DOM events or timeline-based, such as [view progress timelines]().
>
>This module extends the [CSS animations module](), defining triggered animations and methods of controlling {{@keyframe}} animation playback, and uses the timeline and ranges concept introduced in the [CSS scroll-driven animations module]().

from the spec:

[timeline triggers](https://drafts.csswg.org/animation-triggers-1/#timeline-trigger), managed by the [timeline-trigger](https://drafts.csswg.org/animation-triggers-1/#propdef-timeline-trigger) properties, which allow animations to be triggered by entering or leaving certain timeline ranges. (Usually, [view progress timelines](https://drafts.csswg.org/scroll-animations-1/#view-progress-timelines), so an animation can be started when an element comes on-screen.)

Author: Right, I've added a bit more information at the start of the module page, but I think that's enough.

> The **CSS animation triggers** module provides functionality for triggering standard time-based [CSS animations](/en-US/docs/Web/CSS/Guides/Animations) when a particular trigger occurs, such as the user scrolling an element to a certain scroll offset inside the viewport, or a JavaScript event firing on an element.

Reviewer: ```suggestion
The **CSS animation triggers** module provides functionality for triggering standard time-based [CSS animations](/en-US/docs/Web/CSS/Guides/Animations) when a particular trigger occurs, such as the user scrolling an element to a certain scroll offset inside the viewport.
```
let's leave this off until supported (without it is not wrong. with it will lead to questions)

## Animation triggers in action

> Scroll down and up the box of text. The appearance of the "bouncer" text will trigger the ball's animation. The "bouncer" text scrolling out of the viewport triggers a reverse animation. Keep scrolling: the same process will repeat when the "another bouncer" text appears in and then disappears from the viewport.

Reviewer: ```suggestion
This example demonstrates [scroll-triggered animations](/en-US/docs/Web/CSS/Guides/Animation_triggers/Using_scroll-triggered_animations). Scroll the box up and down — the appearance of the "bouncer" text will trigger the ball's animation, and its disappearance will trigger the same animation in reverse. Scroll further, and the same process will repeat when the "another bouncer" text appears and then disappears from the scrollport.
```

> This example demonstrates [scroll-triggered animations](/en-US/docs/Web/CSS/Guides/Animation_triggers/Using_scroll-triggered_animations). Scroll the box up and down — the appearance of the "bouncer" text will trigger the ball's animation, and its disappearance will trigger the same animation in reverse. Scroll further, and the same process will repeat when the "another bouncer" text appears and then disappears from the scrollport.

Reviewer: define what a scroll-triggered animation is in the intro before this.

Author: It's defined in the page intro above.

    height: 5vw;
    display: block;
    aspect-ratio: 1/1;
  }

  @keyframes moveright {
    from {
      transform: translatex(0);
    }
    to {
      transform: translatex(90vw);
    }
  }

  @keyframes bounce {
    9%,
    24%,
    35%,
    44%,
    51%,
    58%,
    63%,
    68%,
    72%,
    76%,
    to {
      transform: translatey(0);
      animation-timing-function: ease-out;
    }
    from,
    17%,
    30%,
    40%,
    48%,
    55%,
    61%,
    66%,
    70%,
    74% {
      animation-timing-function: ease-in;
    }
    0% {
      transform: translatey(-96vh);
    }
    17% {
      transform: translatey(-57.6vh);
    }
    30% {
      transform: translatey(-34.56vh);
    }
    40% {
      transform: translatey(-20.74vh);
    }
    48% {
      transform: translatey(-12.44vh);
    }
    55% {
      transform: translatey(-7.46vh);
    }
    61% {
      transform: translatey(-4.48vh);
    }
    66% {
      transform: translatey(-2.69vh);
    }
    70% {
      transform: translatey(-1.61vh);
    }
    74% {
      transform: translatey(-0.97vh);
    }
  }
}
```

```css hidden live-sample___in-action
@supports not (timeline-trigger-name: --t) {
  body::before {

>     content: "Your browser does not support scroll-triggered animations.";

Reviewer: ```suggestion
    content: "Your browser does not support CSS animation triggers.";
```
because we're on the animation triggers page, but likely JS animation triggers work ;)

Author: I don't think this needs changing, given that we just changed the paragraph preceding the example to say that it demonstrates scroll-triggered animations.

Reviewer: This is the module page, and it's a "see this in action" rather than a code explanation, so many people will not be reading the explanation. "In action" examples are quick examples just to show the module, not a particular feature within the module, which is why the change is needed.

Author: Well, I disagree.

Reviewer: Keeping is ok if we've defined "scroll-triggered animation".

Reviewer 2: Because the check here checks for `timeline-trigger-name`, the wording “scroll-triggered animations” is better than “CSS animation triggers”, as the latter is too broad.

Technically speaking though, to match what the check does, the wording should be “Timeline Triggers” as there can be various types of timelines (if not now, could be in the future).

To keep the original wording (which is “scroll-triggered animations”) I suggest to change the check to the following, which does an actual check for a scroll-timeline being the source:

```css
@supports not (timeline-trigger-source: view()) {
  …
}
```

Author: OK, I've:

1. changed the support query to `timeline-trigger-source: view()`.
2. Added a bit of wording to the intro paragraph, as we define scroll-triggered animations there:

> The **CSS animation triggers** module provides functionality for triggering standard time-based [CSS animations](/en-US/docs/Web/CSS/Guides/Animations) when a particular trigger occurs, such as the user scrolling an element to a certain scroll offset inside the scrollport (this is called a **scroll-triggered animation**).

    background-color: wheat;
    padding: 1rem 0;
    text-align: center;
    padding: 1rem 0;

    z-index: 1;
    position: fixed;
    inset: 40% 0 auto;
  }
}
```

{{embedlivesample("in-action", "100%", 400)}}

> Both trigger elements have a {{cssxref("timeline-trigger-name")}} and {{cssxref("timeline-trigger-source")}}, defining the text blocks as named scroll-view animation triggers. The bouncing ball has the bouncing {{cssxref("animation")}} set on it twice, plus an {{cssxref("animation-trigger")}} property that references both trigger names and specifies animation actions to perform whenever an animation is activated by a trigger coming into view and deactivated when the trigger exits the viewport.

Reviewer: ```suggestion
Both trigger elements have a {{cssxref("timeline-trigger-name")}} and {{cssxref("timeline-trigger-source")}}, defining the text blocks as named scroll-view animation triggers. The bouncing ball has the bouncing {{cssxref("animation")}} set on it twice, plus an {{cssxref("animation-trigger")}} property that references both trigger names and specifies animation actions to perform whenever an animation is activated by a trigger coming into view and deactivated when the trigger exits the scrollport.
```
i'll stop changing viewport to scrollport, and let you do it in bulk

Author: I've just done it in bulk, on every page in the PR.

## Reference

### Properties

- {{cssxref("animation-trigger")}}
- {{cssxref("timeline-trigger")}} shorthand
- {{cssxref("timeline-trigger-activation-range")}} shorthand
  - {{cssxref("timeline-trigger-activation-range-end")}}
  - {{cssxref("timeline-trigger-activation-range-start")}}
- {{cssxref("timeline-trigger-active-range")}} shorthand
  - {{cssxref("timeline-trigger-active-range-end")}}
  - {{cssxref("timeline-trigger-active-range-start")}}
- {{cssxref("timeline-trigger-name")}}
- {{cssxref("timeline-trigger-source")}}
- {{cssxref("trigger-scope")}}

The CSS animation triggers module also introduces the `event-trigger`, `event-trigger-name`, and `event-trigger-source` properties. Currently, no browsers support these features.

### Data types and values

- {{cssxref("&lt;animation-action>")}}

## Guides

- [Using CSS scroll-triggered animations](/en-US/docs/Web/CSS/Guides/Animation_triggers/Using_scroll-triggered_animations)
  - : A guide to implementing CSS scroll-triggered animations.
- [Timeline range names](/en-US/docs/Web/CSS/Guides/Scroll-driven_animations/Timeline_range_names)

>   - : The {{cssxref("timeline-range-name")}} data type: Understanding the various timeline range names.

Reviewer: Should we add the offsets guide too?

Author: Sure. Done.

---
title: CSS animation triggers
short-title: Animation triggers
slug: Web/CSS/Guides/Animation_triggers
page-type: css-module
spec-urls: https://drafts.csswg.org/animation-triggers-1/
sidebar: cssref
---

The **CSS animation triggers** module provides functionality for triggering standard time-based [CSS animations](/en-US/docs/Web/CSS/Guides/Animations) when a particular timeline trigger occurs, such as a scrolling element entering or leaving certain timeline ranges (this is called a **scroll-triggered animation**).

> 

Reviewer: Let's explain, in detailed lay terms, what this module does. How about something like:

```suggestion
The **CSS animation triggers** module provides functionality for **scroll-triggered animation**, standard time-based [CSS animations](/en-US/docs/Web/CSS/Guides/Animations) that are initiated when a particular timeline trigger occurs, letting you control when a regular, {{domxref("DocumentTimeline")}}-based{{cssxref("@keyframes")}} animation starts, pauses, or resets based on the progress of triggering elements' timeline.

With animation triggers, instead of running automatically, a regular time-based keyframe animation can be a _triggered animation_, that begins animating only when an _animation trigger_ activates, such as a scroll‑driven range coming into view. These  _timeline triggers_ generally activate when the trigger element enters or leaves a specified [range](/en-US/docs/Web/CSS/Guides/Scroll-driven_animations/Timeline_insets) (commonly using [view progress timelines](/en-US/docs/Web/CSS/Guides/Scroll-driven_animations/Timelines#view_progress_timelines)). Triggers can specify different actions when entering versus exiting the trigger range, allowing control over animation playback. 

Scroll triggered animations make it possible to effectively delay an animation until it is triggered based on it or another element's scrolling or visibility, without relying on JavaScript.

The animation trigger module also defines **event triggers**. When supported, these will be used to activate timeline-based animations when specific DOM events occur.
```

Author: OK, I've added a bit more detail based on your suggestion here.

---
title: CSS animation triggers
short-title: Animation triggers
slug: Web/CSS/Guides/Animation_triggers
page-type: css-module
spec-urls: https://drafts.csswg.org/animation-triggers-1/
sidebar: cssref
---

The **CSS animation triggers** module provides functionality for triggering standard time-based [CSS animations](/en-US/docs/Web/CSS/Guides/Animations) that are initiated when a particular timeline trigger occurs.

> **scroll-triggered animations** allow you to control when a regular time-based animation starts, pauses, or stops based on when a trigger activates or deactivates, without using JavaScript. This could for example include when a scrolling element enters or leaves a [timeline range](/en-US/docs/Web/CSS/Guides/Scroll-driven_animations/Timeline_insets).

Reviewer: ```suggestion
**Scroll-triggered animations** allow you to control when a regular time-based animation starts, pauses, or stops based on when a trigger activates or deactivates, without using JavaScript. This could, for example, include when a scrolling element enters or leaves a [timeline range](/en-US/docs/Web/CSS/Guides/Scroll-driven_animations/Timeline_insets).
```

Author: Perfect, thanks. I've included this in my next commit.

> The source of these ranges is normally [view progress timelines](/en-US/docs/Web/CSS/Guides/Scroll-driven_animations/Timelines#view_progress_timelines), enabling, for example, an animation to be started when an element enters a scrollport and paused when it leaves the scrollport. Triggers can specify different actions when entering versus exiting the timeline range, allowing control over animation playback.

Reviewer: ```suggestion
The source of these ranges is normally [view progress timelines](/en-US/docs/Web/CSS/Guides/Scroll-driven_animations/Timelines#view_progress_timelines), enabling, for example, an animation to be started when an element enters a scrollport and paused when it leaves the scrollport.  Properties enable changing the timeline range, as well as controlling the active and activation ranges. Triggers can specify different actions when entering versus exiting the timeline range, allowing control over animation playback.
```

Author: added

## Reference

### Properties

- {{cssxref("animation-trigger")}}
- {{cssxref("timeline-trigger")}} shorthand
- {{cssxref("timeline-trigger-activation-range")}} shorthand
  - {{cssxref("timeline-trigger-activation-range-end")}}
  - {{cssxref("timeline-trigger-activation-range-start")}}
- {{cssxref("timeline-trigger-active-range")}} shorthand
  - {{cssxref("timeline-trigger-active-range-end")}}
  - {{cssxref("timeline-trigger-active-range-start")}}
- {{cssxref("timeline-trigger-name")}}
- {{cssxref("timeline-trigger-source")}}
- {{cssxref("trigger-scope")}}

The CSS animation triggers module level 1 also introduces the `event-trigger`, `event-trigger-name`, and `event-trigger-source` properties. Currently, no browsers support these features.

### Data types and values

- {{cssxref("&lt;animation-action>")}}

## Guides

- [Using CSS scroll-triggered animations](/en-US/docs/Web/CSS/Guides/Animation_triggers/Using_scroll-triggered_animations)

>   - : A guide to implementing CSS scroll-triggered animations.

Author: added

  - : A guide to implementing CSS scroll-triggered animations.
- [Timeline range names](/en-US/docs/Web/CSS/Guides/Scroll-driven_animations/Timeline_range_names)
  - : The {{cssxref("timeline-range-name")}} data type: Understanding the various timeline range names.

## Related concepts

- [CSS animations](/en-US/docs/Web/CSS/Guides/Animations) module
  - {{cssxref("animation-timeline")}}
  - {{cssxref("@keyframes")}} at-rule
  - [`<keyframe-selector>`](/en-US/docs/Web/CSS/Reference/Selectors/Keyframe_selectors)
- [CSS scroll-driven animations](/en-US/docs/Web/CSS/Guides/Scroll-driven_animations) module
  - {{cssxref("animation-range")}} shorthand
    - {{cssxref("animation-range-end")}}
    - {{cssxref("animation-range-start")}}
  - {{cssxref("scroll-timeline")}} shorthand
    - {{cssxref("scroll-timeline-axis")}}
    - {{cssxref("scroll-timeline-name")}}
  - {{cssxref("timeline-scope")}}
  - {{cssxref("view-timeline")}} shorthand
    - {{cssxref("view-timeline-axis")}}
    - {{cssxref("view-timeline-inset")}}
    - {{cssxref("view-timeline-name")}}

## Specifications

{{Specifications}}

## See also

> - [CSS scroll-triggered animations are coming!](https://developer.chrome.com/blog/scroll-triggered-animations) on developer.chrome.com (2025)

Reviewer: add the animation and scroll driven animation guides

Author: Added
